### PR TITLE
JSTL - instructor results by RGQ

### PR DESCRIPTION
**Pull request by kanghj**
Ongoing. @yj-soh, have a look? I am definitely open to rewriting the logic.

[edit] comments are not included yet

### CODE REVIEW

**Comment by kanghj** [comment]
@yj-soh, shall I just merge this into the main results page branch first? You can review it there

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 1]
```diff
@@ -466,7 +670,7 @@ private void finaliseBuildingSectionPanel(
             case GIVER_QUESTION_RECIPIENT:
             case RECIPIENT_QUESTION_GIVER:
                 prepareHeadersForTeamPanelsInSectionPanel(sectionPanel);
-                if (responsesGroupedByTeam != null && !responsesGroupedByTeam.isEmpty()) {
+                if (!responsesGroupedByTeam.isEmpty()) {
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Is `responsesGroupedByTeam` guaranteed non-null now?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 2]
```diff
@@ -475,6 +679,12 @@ private void finaliseBuildingSectionPanel(
                                                                                                 : sectionName);
                 sectionPanel.setDisplayingTeamStatistics(true);
                 break;
+            case RECIPIENT_GIVER_QUESTION:
+                sectionPanel.setSectionName(sectionName);
+                sectionPanel.setSectionNameForDisplay(sectionName.equals(Const.DEFAULT_SECTION) ? "Not in a section"
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Wasn't there a constant for this?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 3]
```diff
@@ -82,6 +86,11 @@ public String toString() {
         public boolean isFirstGroupedByGiver() {
             return this == GIVER_QUESTION_RECIPIENT;
         }
+        
+        public boolean isGroupedBySecondaryParticipant() {
+            return this == RECIPIENT_GIVER_QUESTION;
+        }
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Is there a better name for this?
Maybe something more specific signifying what it looks in the frontend?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 4]
```diff
@@ -189,26 +199,50 @@ public void initForViewByRecipientQuestionGiver(InstructorAttributes instructor,
...
+        
+        // Note that if the page needs to load by ajax, then responses will be empty too,
+        // therefore the check whether the bundle needs to come before this
+        if (bundle.responses.isEmpty()) {
+            // no responses, nothing to initialize
+            return;
+        }
+        
+        setShouldCollapsed(bundle.responses.size() > RESPONSE_LIMIT_FOR_COLLAPSING_PANEL);
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Would it be a good idea to put these three segments into a function returning a boolean (to terminate or not)?
Seems to be used multiple times.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 5]
```diff
@@ -396,16 +424,186 @@ private void buildSectionPanelsForRecipientQuestionGiver(
...
+                                    String additionalInfoId, int primaryParticipantIndex,
+                                    Map.Entry<String, Map<String, List<FeedbackResponseAttributes>>> primaryToSecondaryToResponsesMap) {
+        // first build secondary participant panels for the primary participant panel
+        Map<String, List<FeedbackResponseAttributes>> giverToResponsesMap = primaryToSecondaryToResponsesMap.getValue();
+        List<InstructorFeedbackResultsSecondaryParticipantPanelBody> secondaryParticipantPanels 
+                             = buildSecondaryParticipantPanels(
+                                        additionalInfoId, primaryParticipantIndex, giverToResponsesMap);
+        
+        // construct the primary participant panel
+        String primaryParticipantNameWithTeamNameAppended = bundle.appendTeamNameToName(bundle.getNameForEmail(primaryParticipantIdentifier), bundle.getTeamNameForEmail(primaryParticipantIdentifier));
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Either use temporary variables or line breaks, this line is over 200 characters now.

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 6]
```diff
@@ -457,8 +655,14 @@ private InstructorFeedbackResultsGroupByQuestionPanel buildParticipantGroupByQue
         return participantPanel;
     }
     
-    
-    private void finaliseBuildingSectionPanel(
+    private void finaliseBuildingSectionPanelWithoutStats(InstructorFeedbackResultsSectionPanel sectionPanel,
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: `finalize`?

File: src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java [hunk 7]
```diff
@@ -457,8 +655,14 @@ private InstructorFeedbackResultsGroupByQuestionPanel buildParticipantGroupByQue
...
+                                                          String sectionWithoutResponses) {
```
Comment (yj-soh) [inline src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java]: Is it still always a `sectionWithoutResponses`?

File: src/main/webapp/WEB-INF/tags/instructor/results/secondaryParticipantPanel.tag [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+<%@ attribute name="secondaryParticipantPanelBody" type="teammates.ui.template.InstructorFeedbackResultsSecondaryParticipantPanelBody" required="true" %>
+<%@ attribute name="shouldCollapsed" type="java.lang.Boolean" required="true" %>
+<%@ attribute name="showAll" type="java.lang.Boolean" required="true" %>
+<%@ attribute name="primaryParticipantPanel" type="teammates.ui.template.InstructorFeedbackResultsGroupByParticipantPanel" required="true" %>
+<%@ attribute name="secondaryParticipantIndex" type="java.lang.Integer" required="true" %>
+
+
+
+<div class="row ${secondaryParticipantIndex == 0? '': 'border-top-gray'}">
+    <%-- Side of the panel --%>
```
Comment (yj-soh) [inline src/main/webapp/WEB-INF/tags/instructor/results/secondaryParticipantPanel.tag]: Should this be extracted into its own tag if it even requires a comment mentioning how long it stretches for?

File: src/main/webapp/WEB-INF/tags/instructor/results/teamPanel.tag [hunk 9]
```diff
@@ -22,7 +23,15 @@
         <div class="inline panel-heading-text">
             <strong>${teamName}</strong>                        
         </div>
-        <span class="glyphicon ${!shouldCollapsed ? 'glyphicon-chevron-up' : 'glyphicon-chevron-down'} pull-right"></span>
+        <div class="pull-right">
+            <c:if test="${!isDisplayingTeamStatistics}">
+                <a class="btn btn-warning btn-xs" id="collapse-panels-button-team-${teamIndex}" data-toggle="tooltip" title="Collapse or expand all student panels. You can also click on the panel heading to toggle each one individually.">
```
Comment (yj-soh) [inline src/main/webapp/WEB-INF/tags/instructor/results/teamPanel.tag]: Should the variable be named `isDisplayingTeamStatistics`?
I can't really tell why should this happen if `!isDisplayingTeamStatistics`.
Maybe something related to the view instead?

File: src/test/resources/pages/instructorFeedbackResultsSortSecondSessionRecipientGiverQuestion.html [hunk 10]
```diff
@@ -463,19 +460,18 @@
...
-                                        <a data-original-title="Collapse or expand all student panels. You can also click on the panel heading to toggle each one individually." class="btn btn-warning btn-xs" id="collapse-panels-button-team-2" data-toggle="tooltip" title="">
-                                            Collapse  Students
-                                        </a>
-                                        &nbsp;
-                                        <span class="glyphicon glyphicon-chevron-up"></span>
-                                    </div>
-                                </div>
+                            <div class="inline panel-heading-text">
+                        <strong>
+                           Anonymous student 227201833&#39;s Team
```
Comment (yj-soh) [inline src/test/resources/pages/instructorFeedbackResultsSortSecondSessionRecipientGiverQuestion.html]: Is this supposed to be escaped?

File: src/main/webapp/WEB-INF/tags/instructor/results/teamPanel.tag (hunk 9 above)
Comment (kanghj) [inline src/main/webapp/WEB-INF/tags/instructor/results/teamPanel.tag]: If the statistics are shown, then the "Collapse Student/Team" button is displayed below the team statistics table instead. I'll add a comment

**Comment by kanghj** [comment]
@yj-soh, I'll close this branch, and make a new branch containing both RGQ and GRQ.

**Pull request closed without merging**